### PR TITLE
Fix unsafe.string causing some stored values to be changed

### PR DESCRIPTION
**Pull request by CJKkkk-315**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue # #17337

## What this PR does / why we need it:

The key type of the plan cache is converted from an unsafe string to a hash value to prevent modification.
PrepareStmt.ParamTypes uses make to allocate new memory to prevent modification.

___

### **PR Type**
Bug fix

___

### **Description**
- Fixed memory allocation issue for `PrepareStmt.ParamTypes` by using `make` and `copy` to prevent modification.
- Converted SQL strings to SHA-256 hash values for plan cache keys to ensure safety and prevent modification.
- Added a new method `hashString` to generate SHA-256 hash values.
- Updated cache methods (`cache`, `get`, `isCached`) to use hashed keys instead of raw SQL strings.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mysql_protocol.go</strong><dd><code>Fix memory allocation for PrepareStmt.ParamTypes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/frontend/mysql_protocol.go

<li>Added <code>paramTypes</code> variable to store parameter types temporarily.<br> <li> Used <code>make</code> and <code>copy</code> to allocate new memory for <code>stmt.ParamTypes</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17338/files#diff-df9bbc34115212918609e47ad31cf72bf7b2be07d21c34c045074def608d0462">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>plan_cache.go</strong><dd><code>Use SHA-256 hash for plan cache keys</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/frontend/plan_cache.go

<li>Converted SQL strings to SHA-256 hash values for cache keys.<br> <li> Added <code>hashString</code> method to generate hash values.<br> <li> Updated cache methods to use hashed keys.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/17338/files#diff-eba0f4e3a5d0b2011d8a4d0dc72121c6214330b094bb4cbe799151035e0384b5">+15/-5</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: pkg/frontend/plan_cache.go [hunk 1]
```diff
@@ -104,3 +108,9 @@ func (pc *planCache) clean() {
 	pc.lruList = nil
 	pc.cachePool = nil
 }
+
+func (pc *planCache) hashString(sql string) string {
+	hash := sha256.New()
+	hash.Write([]byte(sql))
```
Comment (daviszhen) [inline pkg/frontend/plan_cache.go:114]: 在第一次取unsafe string的地方，就把hash 值算好。放到userInput 中。这里就不用在从string 转 bytes 来一遍了。

File: pkg/frontend/mysql_protocol.go [hunk 2]
```diff
@@ -806,7 +807,9 @@ func (mp *MysqlProtocolImpl) ParseExecuteData(ctx context.Context, proc *process
 
 			// Just the first StmtExecute packet contain parameters type,
 			// we need save it for further use.
-			stmt.ParamTypes, pos, ok = mp.readCountOfBytes(data, pos, numParams<<1)
+			paramTypes, pos, ok = mp.readCountOfBytes(data, pos, numParams<<1)
+			stmt.ParamTypes = make([]byte, len(paramTypes))
```
Comment (daviszhen) [inline pkg/frontend/mysql_protocol.go:811]: func (prepareStmt *PrepareStmt) Close()

里面将paramtType = nil

**Review by daviszhen — CHANGES_REQUESTED** [review]

File: pkg/frontend/plan_cache.go (hunk 1 above)
Comment (CJKkkk-315) [reply pkg/frontend/plan_cache.go:114]: userInput在很多地方有初始化，修改起来有点乱，不如改成这样可以吗，避免了进行额外的内存分配
func (pc *planCache) hashString(sql string) string {
	hash := sha256.New()
	hash.Write(util.UnsafeStringToBytes(sql))
	return hex.EncodeToString(hash.Sum(nil))
}

Comment (daviszhen) [reply pkg/frontend/plan_cache.go:114]: hashString  不用放到planCache下面吧。做成一个工具函数即可。

File: pkg/frontend/mysql_cmd_executor.go [hunk 3]
```diff
@@ -2708,6 +2708,7 @@ func doComQuery(ses *Session, execCtx *ExecCtx, input *UserInput) (retErr error)
 	proc.Base.SessionInfo.User = userNameOnly
 	proc.Base.SessionInfo.QueryId = ses.getQueryId(input.isInternal())
 
+	input.hashedSql = hashString(input.getSql())
```
Comment (daviszhen) [inline pkg/frontend/mysql_cmd_executor.go:2711]: 1,放到上面的 ses.SetSql(input.getSql()) 后面
2, 改为 input.genHash()

File: pkg/frontend/mysql_cmd_executor.go [hunk 4]
```diff
@@ -2841,7 +2842,7 @@ func doComQuery(ses *Session, execCtx *ExecCtx, input *UserInput) (retErr error)
 
 	} // end of for
 
-	if canCache && !ses.isCached(input.getSql()) {
+	if canCache && !ses.isCached(input.getHashedSql()) {
```
Comment (daviszhen) [inline pkg/frontend/mysql_cmd_executor.go:2845]: 改为input.getHash()

**Review by daviszhen — CHANGES_REQUESTED** [review]

**Review by daviszhen — APPROVED** [review]

**Pull request merged**